File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/SimpleSerializer.java
package it.unibo.coordination.linda.presentation;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;
import java.util.Collection;
import java.util.Objects;

public class SimpleSerializer<T> implements Serializer<T> {

    private final MIMETypes mimeType;
    private final ObjectMapper mapper;

    public SimpleSerializer(MIMETypes mimeType, ObjectMapper mapper) {
        this.mimeType = Objects.requireNonNull(mimeType);
        this.mapper = Objects.requireNonNull(mapper);
    }

    @Override
    public MIMETypes getSupportedMIMEType() {
        return mimeType;
    }

    @Override
    public Object toDynamicObject(T object) {
        throw new UnsupportedOperationException();
    }

    @Override
    public String toString(T object) {
        final StringWriter sw = new StringWriter();
        write(object, sw);
        return sw.toString();
    }

    @Override
    public String toString(Collection<? extends T> objects) {
        final StringWriter sw = new StringWriter();
        write(objects, sw);
        return sw.toString();
    }

    @Override
    public void write(T object, Writer writer) {
        writeImpl(object, writer);
    }

    protected void writeImpl(Object object, Writer writer) {
        try {
            mapper.writeValue(writer, object);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Cannot convert " + object + " into " + mimeType, e);
        } catch (IOException e) {
            throw new IllegalStateException("Cannot convert " + object + " into " + mimeType, e);
        }
    }

    @Override
    public void write(Collection<? extends T> objects, Writer writer) {
        writeImpl(objects, writer);
    }

    protected final ObjectMapper getMapper() {
        return mapper;
    }

    public final <X> Serializer<X> getSerializer(Class<X> klass) {
        return Presentation.getSerializer(klass, mimeType);
    }
}

File: presentation/src/main/kotlin/it/unibo/presentation/SimpleSerializer.kt
package it.unibo.presentation

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import java.io.IOException
import java.io.StringWriter
import java.io.Writer

open class SimpleSerializer<T>(override val supportedMIMEType: MIMETypes, protected val mapper: ObjectMapper) : Serializer<T> {

    override fun toDynamicObject(`object`: T): Any {
        throw UnsupportedOperationException()
    }

    override fun toString(`object`: T): String {
        val sw = StringWriter()
        write(`object`, sw)
        return sw.toString()
    }

    override fun toString(objects: Collection<T>): String {
        val sw = StringWriter()
        write(objects, sw)
        return sw.toString()
    }

    override fun write(`object`: T, writer: Writer) {
        writeImpl(`object` as Any, writer)
    }

    protected fun writeImpl(`object`: Any, writer: Writer?) {
        try {
            mapper.writeValue(writer, `object`)
        } catch (e: JsonProcessingException) {
            throw IllegalArgumentException("Cannot convert " + `object` + " into " + supportedMIMEType, e)
        } catch (e: IOException) {
            throw IllegalStateException("Cannot convert " + `object` + " into " + supportedMIMEType, e)
        }
    }

    override fun write(objects: Collection<T>, writer: Writer) {
        writeImpl(objects, writer)
    }

}